### DISCUSSION TITLE
List

### DISCUSSION THREAD

**Discussion by BLoHny** [subak-ba/cs-study#24] [자료 구조]
# List

## 개요
- ADT의 관점에서 Value를 저장하는 추상적인 자료형
- 중복 허용
- 순서가 있음

### ArrayList (배열을 활용해 리스트를 구현)
<img src="https://codehs.gitbooks.io/apjava/content/static/dataStructures/Data_Structures_ArrayList_Example.gif">

### ArrayList의 연산
ArrayList에는 연산이 삽입, 삭제, 데이터확인, 쉬프트(자동) 등이 있습니다.
데이터 확인과 삽입에서는 시간복잡도가O(1) 걸리지만
add(index, n)일 경우와, 나머지 연산은 O(n)이 걸립니다. 
또한 add(index, n)에서는 경우의 수와 삽입 위치에 따라 시간복잡도가 나눠집니다.

#### 삽입

ArrrayList에서의 삽입은 add(n)과, add(index, n) 두가지가 존재합니다.
- add(n)은 현재 배열의 맨 뒤쪽에 데이터를 삽입하는것이고, add(index, n은) 그 인덱스에 데이터를 삽입하는 것 입니다.
![image](https://github.com/subak-ba/cs-study/assets/113517233/79943e46-365d-459c-9d92-d57693aa3e4f)

- add(index, n))을 수행할때는 그 인덱스에 이미 데이터가 존재할 경우 그 인덱스부터 뒤의 데이터까지를 오른쪽으로 shift하고 삽입하게되므로 추가적인 시간이 걸리게됩니다.

![image](https://github.com/subak-ba/cs-study/assets/113517233/bd207d84-f00e-4c17-bf86-12858d0070ec)

#### 삭제

ArrayList에서의 삭제는 remove(n) 과 remove(index)가 있습니다.
- remove(n)은 가장 첫 인덱스부터 배열의 마지막까지 순회하며 가장첫 contains(n)인 인덱스를 지우고 시프트를 수행해야 한다면 수행합니다.
![image](https://github.com/subak-ba/cs-study/assets/113517233/2103d273-8849-4500-aa04-e14b54be5034)

- remove(index)는 인덱스번호의 데이터값을 지우고 시프트를 수행해야 한다면 수행합니다.
![image](https://github.com/subak-ba/cs-study/assets/113517233/e5fc6112-1d23-4731-8b2a-1d45b728e28e)

#### 데이터 확인

ArrayList에서의 데이터확인은 contains(n)과 get(index) 가 있습니다.
- contains(n)은 부여한 n값이 ArryaList에 존재한다면 true 아니라면 false를 반환합니다.
![image](https://github.com/subak-ba/cs-study/assets/113517233/a8fb5e0e-38a9-4fc2-b029-7dc736b7493d)

- get(index)는 index번호의 값을 꺼내옵니다.
![image](https://github.com/subak-ba/cs-study/assets/113517233/61fb1898-b2e3-4abc-ad0d-537ef1f6356c)

---

### LinkedList (연결 리스트)
LinkedList에는 ArrayList와 거의 똑같이 연산이 삽입, 삭제, 데이터확인 등이 있습니다.
다만 차이점으로는 LinkedList는 index(0)의 Head와 index(length) 의 Tail을 기록하고 있습니다.
또한 시간복잡도에서는 데이터 접근과 탐색에서 O(n)이 걸리지만,
쉬프트 연산이 존재하지않고 단순이 기존 index의 노드 연결부분만 이어주면 되므로 삽입/삭제는 o(1)이 걸립니다.

- add(n), add(index, n)
  - add(n)의 맨뒤로 삽입하게되고 tail을 이어받습니다.
  - add(index, n)의 경우 그 index가 head나 tail일 경우 이어받고 기존 head, tail노드와 연결하게 됩니다.
![image](https://github.com/subak-ba/cs-study/assets/113517233/c32c5cb7-5f04-4407-8f13-03d0f8ae37f2)

- remove(n), remove(index, n)
  - remove(n)은 head부터 데이터를 순회하며 가장먼저 만난값을 삭제후 next노드와 자신의 앞 노드를 연결합니다. 삭제한 node가 tail일 경우 tail을 땅기게 됩니다.
  - remove(index, n) 또한 head=0 을 기준으로 순회하다 해당 index의 값을 삭제후 next노드와 자신의 앞 노드를 연결합니다. 삭제한 node가 tail일 경우 tail을 땅기게 됩니다.

- get(index), contains(n)
  - get(index)는 head=0을 기준으로 순회하다 해당 index의 값을 반환합니다.
  - contains(n)은 head를 기준으로 데이터를 순회하며 해당 데이터가 있을경우 true 아닐경우 false를 반환합니다.

## LinkedList의 사촌들
- Circular linked list
Head가 존재하지 않고 tail과 index(0)의 노드가 연결되있는것 (기존에는 tail은 null을 가리킴)
![image](https://github.com/subak-ba/cs-study/assets/113517233/07c179a6-3039-4574-9dfd-7e84a8b8f672)

- Doubly linked list
기존에 노드가 자신의 next 만 기록하던 것과 별게로 자신의 앞선 노드와도 연결하게 되어 탐색이 쉬어집니다.
![image](https://github.com/subak-ba/cs-study/assets/113517233/386c74bc-0bbe-457d-a4d3-20a6d99d9f2d)

- Circular doubly linked list
Circular linked list와 Doubly linked list를 합친 리스트로 prev, tail -> head를 모두 가지고있다
![image](https://github.com/subak-ba/cs-study/assets/113517233/e5144c8f-c718-43ed-bfaa-54bcc2d3bf96)

## ArrayList vs LinkedList
![image](https://github.com/subak-ba/cs-study/assets/113517233/eb02b79a-8777-40e0-84e5-f3c78fac1307)

단순히 처음에 글을 쓸때는 쓸데없는 메모리가 생길수도있는 ArrayList가 LinkedList에 비해 오버헤드가 있을 수도 있지 않을까라는 생각이 들었지만 ArrayList는 CPU Cache를 사용할 수 있다는 점과 ArrayList의 배열복사나 쉬프트연산이 계속된 튜닝을 통해 하나하나 하는것이 아닌 속도가 개선되어왔다는점에서 ArrayList를 사용해도 괜찮다는 생각이 들었습니다.

### 여담
또한 Java기준 List를 작성한 Joshua Bloch의 twitter 글을 참조하면
![image](https://github.com/subak-ba/cs-study/assets/113517233/84d8c835-5021-4e52-a8a6-bbd7a92e99b3)

내가 LinkedList를 만들었지만 나는 사용해 본 적이 없다는 웃긴 답변또한 있다
트위터글을 좀 더 본 결과 FIFO(First in First output)을 위해 LinkedList를 사용한다 라는 댓글에는 ArrayDeque라는 좀 더 성능이 좋은 클래스가 있으니 그걸 사용하라는 댓글 또한 보았다

### List를 구현한 다른 아이들

- Vector 
  - ArryaList와 다르게 thread-safe한 list 구조
  - 하지만 Vector를 구현할 이유가 굳이 없을때는 메모리 절약상 ArrayList를 사용하는것이 좋다 Vector는 크기확장시 배열을 2배늘리는 반면 ArrayList는 50%만 늘리기 때문
- Stack
  - Vector를 확장한 LIFO 구조
- Queue
  - FIFO 구조

## 코드 구현
- ArrayList
```java 
package list;

public class ArrayListed {
    
    private Integer[] elements;
    private int size;

    public ArrayListed(int size) {
        elements = new Integer[size]; 
        this.size = 0;
    }

    //add(n)
    public void add(int a) {
        if (size == elements.length) {
            copyArray(); 
        }
        elements[size++] = a;
    }

    //get(index)
    public int get(int index) {
        if (index >= size || index < 0) {
            throw new IndexOutOfBoundsException();
        }
        return elements[index];
    }

    //remove(n)
    public void removeByValue(int a) {
        for (int i = 0; i < size; i++) {
            if (elements[i].equals(a)) {
                // 요소를 앞으로 이동시키는 로직
                System.arraycopy(elements, i + 1, elements, i, size - i - 1);
                elements[--size] = null; // 마지막 위치를 null로 설정하여 객체 참조 제거
                break; 
            }
        }
    }

    // 배열 크기 증가
    private void copyArray() {
        int newSize = elements.length * 2;
        Integer[] newElements = new Integer[newSize];
        for (int i = 0; i < size; i++) {
            newElements[i] = elements[i];
        }
        elements = newElements;
    }
}
```
- LinkedList
```java
public class LinkedListed {
    private Node head;

    public LinkedListed() {
        this.head = null;
    }

    public Integer getIndex(int index) {
        Node current = head;
        for (int i = 0; i < index; i++) {
            current = current.next;
        }
        return current.data;
    }

    public void addHead(Integer data) {
        Node newNode = new Node(data);
        newNode.next = head;
        head = newNode;
    }

    public void addTail(Integer data) {
        Node newNode = new Node(data);
        if (head == null) {
            head = newNode;
            return;
        }
        Node last = head;
        while (last.next != null) {
            last = last.next;
        }
        last.next = newNode;
    }

    public boolean remove(Integer value) {
        if (head == null) return false;

        if (head.data.equals(value)) {
            head = head.next;
            return true;
        }

        Node current = head;
        while (current.next != null) {
            if (current.next.data.equals(value)) {
                current.next = current.next.next;
                return true;
            }
            current = current.next;
        }

        return false;
    }
}

class Node {
    Integer data;
    Node next;

    public Node(Integer data) {
        this.data = data;
        this.next = null;
    }
}
```

#18

**Comment by onee-only** [comment]
왜 연결 리스트는 배열로 만든 리스트만큼 많이 사용되지 않을까요?
배열과 CPU L-캐시가 어떤 연관이 있기에 연결 리스트가 배열로 만든 리스트보다 느릴까요?

[이 글](https://www.reddit.com/r/learnprogramming/comments/10hwrmy/why_arent_linked_lists_more_popular/)을 참고해 보시면 좋을 것 같습니다.
